### ISSUE TITLE
Electrostatics direct-space/reciprocal-space transition when NonbondedForce method is PME?

### ISSUE THREAD

**Issue by j-wags** [openmm/openmm#3103]
In the following code, is the transition between direct-space and reciprocal-space for the PME electrostatics set by the `cutoff` variable, or is the `cutoff` only used for LJ (and the electrostatics PME transition is determined automatically by the error tolerance)?

```
force = openmm.NonbondedForce()
force.setNonbondedMethod(openmm.NonbondedForce.PME)
force.setUseDispersionCorrection(True)
force.setCutoffDistance(cutoff)
```

CC https://github.com/openforcefield/openff-toolkit/pull/917#issuecomment-831618918

**Comment by peastman** [comment]
See http://docs.openmm.org/latest/userguide/theory.html#coulomb-interaction-with-particle-mesh-ewald.  It describes exactly how the PME parameters get calculated from the cutoff and error tolerance.

**Comment by j-wags** [comment]
Thanks! So it looks like r_cutoff _is_ used as the transition between direct- and reciprocal-space, but the number of grid points changes as a function of r_cutoff to achieve the desired error tolerance. So, the choice of r_cutoff does not uniquely describe the resulting PME energy. 

A lot of previous discussions are making sense now -- So would it be correct to say that getting a unique energy for a OpenMM PME-dependent force requires knowing both 
* the platform, and 
* TWO of {r_cutoff, n_grid_points, error_tolerance, alpha}? 

Apologies if this is creating spam, and I'm happy to have this discussion elsewhere or do a deeper dive into the docs before taking more of your time. I'm working on better understanding these factors so that I can be a better steward of the [SMIRNOFF spec](https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html) and incorporate your previous comments on it.

**Comment by peastman** [comment]
It depends on how you specify it.  If you specify the cutoff, error tolerance, and box shape, then alpha and the number of grid points along each axis are calculated from them.  Alternatively you can set the cutoff, number of grid points, and alpha.  In that case the error tolerance is ignored.  Also note that the number of grid points is calculated from the *initial* box shape.  If you have a barostat, the box may change size during the simulation but the number of grid points stays the same.

And of course other things also matter.  The level of numeric precision can be important.  If you're using single precision, trying to making the error tolerance too small can actually make the error larger.  Also note that OpenMM always uses 5th order splines, but some programs let you choose the order.  The heuristic we use for choosing the number of grid points is only appropriate for 5th order splines.  A different order would require a different grid size to get the same accuracy.

You can think of the PME parameters as being underdetermined.  There's no unique "best" set of values.  By changing alpha, you can shift work between the direct space and reciprocal space calculations without changing the accuracy.  By changing the spline order, you can shift work between charge spreading and FFTs without changing the accuracy.  In practice you select an accuracy then try to find the parameters that give the fastest performance while achieving that accuracy.  Since the relative speeds of different parts of the calculation may vary between codes, the optimal settings may be different.

**Comment by j-wags** [comment]
Thanks for the great explanation -- This makes a lot of sense. I'm going to take some time to digest what this means for our force field definition, and will reopen this issue if needed!

I guess one thing we're trying to accomplish is to provide exact instructions for how to combine a molecule and a force field to get a system with a precise energy. But having "use PME" be part of the force field definition leaves a lot undefined. And defining everything required to get a precise PME energy is likely to be super difficult, since details of the PME energy like grid points depend on the box size, which isn't something that's known to the force field, and the heuristics to define the number of grid points can reasonably vary between implementations. 

The bottom line seems to be "it IS possible to specify enough about PME to get deterministic energies out for a particular system+configuration, but it would be so specific that it'd be largely counterproductive, and that's not the point of PME." I'll keep thinking on this, but it seems like there's a case for the use of PME/its settings to be in the "runtime configuration" bucket (like timestep) and not in the FF bucket (like switching width).

**Comment by swails** [comment]
> "it IS possible to specify enough about PME to get deterministic energies out for a particular system+configuration, but it would be so specific that it'd be largely counterproductive, and that's not the point of PME."

It's also worth pointing out that many programs don't implement the same set of options, like spline order or even which parameters you can set.  What seems most sensible to me is to specify something broad but with a well-defined meaning relative to what you're trying to accomplish.  For example, a tolerance (which lets you say something re: the expected level of agreement between different implementations regardless of specific details).

**Comment by peastman** [comment]
And the differences go even deeper than that.  For example, I believe Desmond now uses u-series instead of PME.

My 2 cents on this subject: the force field should define the potential function, but not the algorithm used to calculate it.  For electrostatics, the potential function it specifies is essentially, "Compute all possible interactions between all possible periodic copies of particles, no matter how far apart."  Various algorithms can be used to calculate that quantity: Ewald, PME, u-series, PPPM, etc.  They're all methods for calculating the same quantity, so there's no reason to dictate one of them over the others.

Of course, all of them involve approximations.  It's impossible in general to compute an infinite set of interactions in a finite amount of time!  But all of them are systematically improvable: by tuning the parameters, you can make any one of them arbitrarily accurate.  That's a trade-off between speed and accuracy that the force field also can't dictate.  It's up to the user to decide what levels of speed and accuracy are acceptable for their application.

Also keep in mind that people sometimes simulate non-periodic systems.  In that case, it is possible to truly compute all interactions.  But a user might still choose to use an approximation like FMM to trade accuracy for speed.  The force field shouldn't forbid that.
